feat: Migration and introspection proxy (#421)
electric-sql/electric: components/electric/lib/electric/ddlx/readme.md

# DDLX

There are three distinct bits inside the module Electric.DDLX:

- sql
- parse
- commands

### sql

The SQL statements that install DDLX tables and functions into Postgres. They are actually in a single file in `sql/init_ddlx.sql`.

> You can use this on it own without `parse` or `commands` you want. They are really just utilities to help call the sql.

Reviewer: ```suggestion
You can use this on its own without `parse` or `commands` if you want. They are really just utilities to help call the sql.
```

Reviewer: This remains unaddressed.

Running these statements will:

- install uuid-ossp
- ensure there is an `electric` schema  
- create 3 new DDLX tables
    - `electric.grants`
    - `electric.roles`
    - `electric.assignments`
- add 7 new DDLX functions    
    - `electric.enable`
    - `electric.disable`
    - `electric.assign`
    - `electric.unassign`
    - `electric.grant`
    - `electric.revoke`
    - `electric.sqlite`
- and 3 utility functions that it uses internally
    - `electric.find_pk`
    - `electric.find_fk_for_column`
    - `electric.find_fk_to_table`
    
### commands

These are a set of structs, one for each of the DDLX Postgres functions, that have the common method `pg_sql` which will
return a SQL statment string to call its function.

This can be accessed through `Electric.DDLX.command_to_postgres()`

### parse

These are a set of parsers that read DDLX's extended PostgreSQL language statements and return command structs.

See [https://github.com/electric-sql/docs/blob/main/documentation/docs/api/ddlx.mdx]() for the syntax of DDLX that they read.

This is WIP and not everything works yet.

This can be accessed through `Electric.DDLX.ddlx_to_commands()`.

> These is also an extra utility function `Electric.DDLX.is_ddlx()` to help check if a SQL statement is actuallt a DDLX statement.

Reviewer: ```suggestion
These is also an extra utility function `Electric.DDLX.is_ddlx()` to help check if a SQL statement is actually a DDLX statement.
```

Reviewer: This remains unaddressed.

## Functions internals

### electric.enable

> currently does nothing - is the new 'electrify'

Reviewer: That's not true. It calls `electric.electrify()`.

### electric.disable
currently does nothing - is the new 'unelectrify'

### electric.assign

> Adds and entry to the `electric.assignments` and creates a set of resources used by the assignment:

Reviewer: ```suggestion
Adds a new entry to `electric.assignments` and creates a set of resources used by the assignment:
```

- 1 x join table
- 2 x functions 
- 3 x triggers

> When a table referenced by the assignment is created or modified the triggers run a function that updates both the join table

Reviewer: ```suggestion
When a table referenced by the assignment is created or modified, the triggers run a function that both updates the join table
```

and writes the role into the `electric.roles` table.

> The row in the the join table has foreign keys to:

Reviewer: ```suggestion
The row in the join table has foreign keys to:
```
